File: frontend/src/auth/authService.ts
import { API_BASE } from '../api/client'
import type { Session } from './session'

export type SignInResult =
  | { ok: true; session: Session }
  | { ok: false; message: string }

export async function signIn(
  email: string,
  password: string,
): Promise<SignInResult> {
  let response: Response

  try {
    response = await fetch(`${API_BASE}/api/auth/login`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ email: email.trim(), password }),
    })
  } catch {
    return {
      ok: false,
      message: `Cannot reach the API at ${API_BASE}. Is the backend running?`,
    }
  }

  if (response.status === 401) {
    return { ok: false, message: 'Invalid email or password.' }
  }

  if (!response.ok) {
    return { ok: false, message: `Sign-in failed (HTTP ${response.status}).` }
  }

  return { ok: true, session: (await response.json()) as Session }
}

File: mobile/lib/services/auth_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../config/api_config.dart';
import '../models/auth_models.dart';

class AuthService {
  Future<AuthSession> login({
    required String email,
    required String password,
  }) async {
    final http.Response response;
    try {
      response = await http.post(
        Uri.parse('${ApiConfig.baseUrl}/api/auth/login'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'email': email.trim(), 'password': password}),
      );
    } on http.ClientException {
      throw Exception('Cannot reach the RescueSriLanka server.');
    }

    if (response.statusCode == 200) {
      return AuthSession.fromJson(
        jsonDecode(response.body) as Map<String, dynamic>,
      );
    }

    if (response.statusCode == 401) {
      throw Exception('Invalid email or password.');
    }

    throw Exception('Sign-in failed (HTTP ${response.statusCode}).');
  }
}
